language/vm/vm_runtime/src/code_cache/module_cache.rs

```diff
@@ -249,23 +253,51 @@ impl<'alloc> VMModuleCache<'alloc> {
             SignatureToken::String => Ok(Ok(Some(Type::String))),
             SignatureToken::ByteArray => Ok(Ok(Some(Type::ByteArray))),
             SignatureToken::Address => Ok(Ok(Some(Type::Address))),
-            SignatureToken::TypeParameter(_) => unimplemented!(),
-            SignatureToken::Struct(sh_idx, _) => {
+            SignatureToken::TypeParameter(idx) => Ok(Ok(Some(type_context.get_type(*idx)?))),
+            SignatureToken::Struct(sh_idx, tys) => {
+                let ctx = {
+                    let mut ctx = vec![];
+                    for ty in tys.iter() {
+                        let resolved_type = try_runtime!(self
+                            .resolve_signature_token_with_fetcher(
+                                module,
+                                ty,
+                                type_context,
+                                gas_meter,
+                                fetcher
+                            ));
+                        if let Some(t) = resolved_type {
+                            ctx.push(t);
+                        } else {
+                            return Ok(Ok(None));
+                        }
+                    }
+                    TypeContext::new(ctx)
+                };
                 let struct_def =
                     try_runtime!(self
-                        .resolve_struct_handle_with_fetcher(module, *sh_idx, gas_meter, fetcher));
+                        .resolve_struct_handle_with_fetcher(module, *sh_idx, gas_meter, fetcher))
+                    .map(|def| ctx.subst_struct_def(&def)).transpose()?;
                 Ok(Ok(struct_def.map(Type::Struct)))
             }
             SignatureToken::Reference(sub_tok) => {
-                let inner_ty =
-                    try_runtime!(self
-                        .resolve_signature_token_with_fetcher(module, sub_tok, gas_meter, fetcher));
+                let inner_ty = try_runtime!(self.resolve_signature_token_with_fetcher(
```

Comment:
Some thoughts while relooking at this code. Maybe for the reference case we should just return an error as there should be no reference in a stored struct?